### ISSUE TITLE
Categories not showing in results

### ISSUE THREAD

**Issue by hemorej** [algolia/algoliasearch-magento-2#543]
**Do you want to request a *feature* or report a *bug*?**
Bug

**Bug: What is the current behavior?**
In instant search on Magento 2, categories are not being returned in the response by the APIl

**What is the version of Magento and of Algolia extension you are using? Always use the latest version of the extension one before opening a bug issue.**

Magento (2.2.2), Algolia for vanilla JavaScript (3.24.9), Magento2 integration (1.7.0)

**Details**

- I tried reindexing our site, it makes no difference. I can see the categories in the Algolia dashboard for that site, so it's not an indexing issue
- If I make a UI demo on the same dataset from the Algolia dashboard, with the same query, I get a much different response than from our front end. Notice how the "hits" object is empty even as the UI shows a result. 

**UI FRONTEND:**
Query  
```
{"params":"query: fancy
hitsPerPage: 2
analyticsTags: autocomplete
clickAnalytics: true"}
```

Response  
```
{
   "hits":[],
   "nbHits":0,
   "page":0,
   "nbPages":0,
   "hitsPerPage":2,
   "processingTimeMS":1,
   "exhaustiveNbHits":true,
   "query":"fancy",
   "params":"query=fancy&hitsPerPage=2&analyticsTags=autocomplete&clickAnalytics=true&tagFilters=",
   "queryID":"1e75a59de22d1e8b25f1c2b1cb9ed3c5"
}
```

Same result when using this query ```{"params":"query=fancy&hitsPerPage=10&facets=*&maxValuesPerFacet=10&facetFilters=%5B%5D"}```

**UI DEMO AlGOLIA DASH:**
Query  
```
{"params":"query: fancy
hitsPerPage: 10
facets: *
maxValuesPerFacet: 10
facetFilters: []"}
```

Response  
```
{
   "hits":[
      {
         "name":"Fancy product",
         "url":"https://magento.toonboom.com/fancy-product.html",
         "visibility_search":1,
         "visibility_catalog":1,
         "type_id":"downloadable",
         "ordered_qty":7,
         "rating_summary":"",
         "categories":{
            "level0":[
               "Visual"
            ],
            "level1":[
               "Visual /// Brushes"
            ]
         },
         "categories_without_path":[
            "Brushes"
         ],
         "thumbnail_url":"//magento.toonboom.com/pub/media/catalog/product/cache/fd4c882ce4b945a790b629f572e4ef93/l/8/l8j-urhd.jpeg",
         "image_url":"//magento.toonboom.com/pub/media/catalog/product/cache/0f831c1845fc143d00d6d1ebc49f446a/l/8/l8j-urhd.jpeg",
         "in_stock":1,
         "stock_qty":102,
         "sku":"ABC1234",
         "price":{
            "USD":{
               "default":10,
               "default_formated":"$10.00",
               "special_from_date":false,
               "special_to_date":false
            }
         },
         "created_at":"0000-00-00 00:00:00",
         "algoliaLastUpdateAtCET":"2018-06-11 15:19:54",
         "objectID":"7",
         "_highlightResult":{
            "name":{
               "value":"<em>Fancy</em> product",
               "matchLevel":"full",
               "fullyHighlighted":false,
               "matchedWords":[
                  "fancy"
               ]
            },
            "categories":{
               "level0":[
                  {
                     "value":"Visual",
                     "matchLevel":"none",
                     "matchedWords":[

                     ]
                  }
               ],
               "level1":[
                  {
                     "value":"Visual /// Brushes",
                     "matchLevel":"none",
                     "matchedWords":[

                     ]
                  }
               ]
            },
            "categories_without_path":[
               {
                  "value":"Brushes",
                  "matchLevel":"none",
                  "matchedWords":[

                  ]
               }
            ],
            "sku":{
               "value":"ABC1234",
               "matchLevel":"none",
               "matchedWords":[

               ]
            }
         }
      }
   ],
   "nbHits":1,
   "page":0,
   "nbPages":1,
   "hitsPerPage":10,
   "processingTimeMS":1,
   "facets":{
      "categories.level0":{
         "Visual":1
      },
      "categories.level1":{
         "Visual /// Brushes":1
      },
      "price.USD.default":{
         "10":1
      }
   },
   "facets_stats":{
      "price.USD.default":{
         "min":10,
         "max":10,
         "avg":10,
         "sum":10
      }
   },
   "exhaustiveFacetsCount":true,
   "exhaustiveNbHits":true,
   "query":"fancy",
   "params":"query=fancy&hitsPerPage=10&facets=*&maxValuesPerFacet=10&facetFilters=%5B%5D&restrictIndices=default_products&validUntil=1533910612"
}
```

<img width="969" alt="screen shot 2018-08-10 at 10 03 18" src="https://user-images.githubusercontent.com/5703037/43962595-26d0344a-9c86-11e8-9e4e-cf3ead17eef6.png">

**Comment by JanPetr** [comment]
Hello @hemorej,

categories in autocomplete are not fetched from "products" index, but it's own "categories" index.
Did you run a category indexer to push your categories data to Algolia?

**Comment by hemorej** [comment]
@JanPetr I'm not sure I understand what you mean, I need to search for a product and show the matching category facets on the left in the instant search results box, not a list of all categories. The categories are indexed in Algolia yes

<img width="969" alt="screen shot 2018-08-10 at 10 03 18" src="https://user-images.githubusercontent.com/5703037/44041605-2b6c76e0-9eec-11e8-8965-5c841c3df56a.png">

**Comment by mageprashant** [comment]
I had the same problem (with version 1.7.0) Categories not showing in results
http://nimb.ws/4cbJqx

Thank you

**Comment by aaronhuang1** [comment]
Hi 

I had the similar isssues.

store id 2 (main stoer) showing category, tired reforce index many time.
store id 3 shwoing category 

Please help to fix asap.

![image](https://user-images.githubusercontent.com/17335551/46356740-b0548500-c6a6-11e8-8593-1050ffafe460.png)
![image](https://user-images.githubusercontent.com/17335551/46356841-ea258b80-c6a6-11e8-811d-e28b93f0aa50.png)

**Comment by aaronhuang1** [comment]
@JanPetr ,

any update?? if not we have no choice  to looking for other product to replace Algolia

**Comment by Sil-B** [comment]
Same problems here. Is there any update?

**Comment by evolbug** [comment]
A manual solution is doing a full category reindex manually, for some reason the default indexer does not pick up the categories. I may investigate this deeper later and contribute if it doesn't get fixed by then.
E.g. a shell script I have written:
```php
    /**
     * __construct
     *
     * @param IndexerRegistry $indexerRegistry
     * @param State $state
     */
    public function __construct(IndexerRegistry $indexerRegistry, State $state)
    {
        $this->state = $state;
        $this->indexer = $indexerRegistry->get('algolia_categories');
        parent::__construct();
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $output->writeln("Reindexing categories...");

        $this->state->setAreaCode(Area::AREA_FRONTEND);
        $this->indexer->reindexAll();

        $output->writeln("<info>Success!<info>");
    }
```

**Comment by Cronium** [comment]
Hi evolbug,

Tried your solution, but its not working for me. While it did force one of my storeviews to populate fully, it is still missing a whole store's worth of categories.

**Comment by evolbug** [comment]
@Cronium you could try changing area codes to `AREA_ADMINHTML` or `AREA_GLOBAL`, I haven't tried them as I only worked with one store view. If either has better results, do tell!
Can double check per-view configuration of Algolia just in case as well.

**Comment by evolbug** [comment]
@Cronium 
It is implemented as a Symfony cli command. Here are all the imports 
```
use Magento\Framework\App\Area;
use Magento\Framework\App\State;
use Magento\Framework\Indexer\IndexerRegistry;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
```

**Comment by bayucakra20** [comment]
Hi @evolbug 

I just add command that you mention above, but I've got error while run those command like
```
[Magento\Framework\Exception\LocalizedException]
Area code is already set
```
what is this mean?

**Comment by evolbug** [comment]
@bayucakra20 depending on Magento version area code might already be set, in which case remove `        $this->state->setAreaCode(Area::AREA_FRONTEND);` line

However, if it doesn't work afterwards, I believe you will be out of luck and a permanent fix by the dev is required.

**Comment by Cronium** [comment]
@evolbug 

Hey I've been looking at this some more, could you clarify what the difference between your script, and running php bin/magento index:reindex algolia_categories ?

**Comment by evolbug** [comment]
@Cronium the index:reindex doesn't actually pick up the categories, unless that has changed in a newer version.

**Comment by JanPetr** [comment]
Hello there!

The categories reindexing has been fixed in 1.8.5 version of the extension. However it's not fixed in 1.9.1 version. It'll be included soon. Can you give a try to 1.8.5 version and come back to us with results on your sites?

Thank you!

**Comment by Cronium** [comment]
Any updates on when this fix will be rolled into the master? 1.10 came out 6 days ago and no mention of #543 in the changelogs.

**Comment by JanPetr** [comment]
Hello @Cronium,

sorry, the fix was not included in 1.10.0 :( Right now, it's part of #696, merged to `develop` branch. Can you give it a try? It'll be released with the next major version.

Thank you for your patience 💯
